### PR TITLE
add "healthy" sdnotify policy

### PR DESCRIPTION
**Pull request by vrothberg**
Add a new "healthy" sdnotify policy that instructs Podman to send the READY message once the container has turned healthy.

Fixes: #6160

#### Does this PR introduce a user-facing change?

```release-note
Add a new --sdnotify=healthy policy where Podman sends the READY message once the container turns healthy.
```

For auto: @ygalblum @rhatdan PTAL

### CODE REVIEW

**Review by giuseppe — APPROVED** [review]
LGTM

File: test/system/260-sdnotify.bats [hunk 1]
```diff
@@ -184,6 +184,65 @@ READY=1"
     _stop_socat
 }
 
+# These tests can fail in dev. environment because of SELinux.
+# quick fix: chcon -t container_runtime_exec_t ./bin/podman
+@test "sdnotify : healthy" {
```
Comment (ygalblum) [inline test/system/260-sdnotify.bats:189]: This test checks the behavior when using `health-cmd` which is more a health check than a readiness probe. Consider adding a test for `--health-startup-cmd` which is more suitable to what a user would try to achieve

File: libpod/container_internal.go [hunk 2]
```diff
@@ -1241,7 +1243,32 @@ func (c *Container) start() error {
...
+		return nil
+	}
+
+	// Wait for the container to turn healthy before sending the READY
+	// message.  This implies that we need to unlock and re-lock the
+	// container.
+	if !c.batched {
+		c.lock.Unlock()
+		defer c.lock.Lock()
+	}
```
Comment (Luap99) [inline libpod/container_internal.go:1260]: please update the function comment to reflect that

File: libpod/container_internal.go [hunk 3]
```diff
@@ -1241,7 +1243,32 @@ func (c *Container) start() error {
...
+
+	if _, err := c.WaitForConditionWithInterval(ctx, DefaultWaitInterval, define.HealthCheckHealthy); err != nil {
+		return err
+	}
+
+	if err := notifyproxy.SendMessage(c.config.SdNotifySocket, daemon.SdNotifyReady); err != nil {
```
[the remark this answers is not in the document]
Comment (vrothberg) [reply libpod/container_internal.go:1266]: I think that's OK.

File: libpod/container_internal.go [hunk 4]
```diff
@@ -1241,7 +1243,32 @@ func (c *Container) start() error {
...
+		logrus.Errorf("Sending READY message after turning healthy: %s", err.Error())
```
[the remark this answers is not in the document]
Comment (vrothberg) [reply libpod/container_internal.go:1267]: Yes, I'll update the other caller as well.

File: test/system/260-sdnotify.bats (hunk 1 above)
Comment (vrothberg) [reply test/system/260-sdnotify.bats:189]: Good point.  It seems that `--health-startup-cmd` needs much more testing - it's not tested in `test/system` at all.

I will play a bit with your suggestion but think we need to improve the test coverage even further.

Comment (vrothberg) [reply test/system/260-sdnotify.bats:189]: @ygalblum, can you elaborate on the exact requirements?  IIUIC, the startup "gates" the ordinary healthcheck.  So once the startup HC is good, the ordinary HC is executed.  Only the latter influences the health status.

How do you want it to behave?

**Review by Luap99 — COMMENTED** [review]
LGTM

**Comment by vrothberg** [comment]
/hold

Let's wait for @ygalblum's feedback.

File: test/system/260-sdnotify.bats (hunk 1 above)
Comment (ygalblum) [reply test/system/260-sdnotify.bats:189]: I think I also share your understanding about the order of things in the code. As their name suggests, the startup checks verify that the service has started correctly. Once it has, the health checks kick in to check that the service is in fact still running.

As for my comment, if I look at it from systemd's perspective, it need to know that the service has started which IMHO should correspond with the startup checks more than with the health check.

Having said that, I guess we would like to allow users to define only the health check and let it act as the startup one. But, regardless to whether the health check is set, if the startup health is set, I think that systemd should be notified once it is successful.

Again, from the looks of the code, I think it does that, but I'm not sure. So, we should test it. In addition, as always, tests are there not only to make sure that the code we added it correct, but also to make sure the behavior is not broken unintentionally.

Comment (vrothberg) [reply test/system/260-sdnotify.bats:189]: > tests are there not only to make sure that the code we added it correct, but also to make sure the behavior is not broken unintentionally.

Totally agree.  That's why I am holding the PR.

> As for my comment, if I look at it from systemd's perspective, it need to know that the service has started which IMHO should correspond with the startup checks more than with the health check.

That is currently not possible. The `wait --condition=healthy` logic waits for the main HC, not the startup one.  The latter has no impact on the health status other than it will delay the main HC from running.

So in other words: you are looking for a means to wait for the startup HC to finish?  If that's the case, then I think we can merge this PR as is but also need to take a step back and see how we can cover the startup one.

Comment (vrothberg) [reply test/system/260-sdnotify.bats:189]: Something like `wait --condition=$startup_done` with a nicer name for `$startup_name`.  Once that's done, we can add a `--sdnotify=$startup_done` on top?

Comment (vrothberg) [reply test/system/260-sdnotify.bats:189]: BTW. I really like the idea as it - as you mentions - discriminates the initialization/startup phase from the workload phase.

Comment (ygalblum) [reply test/system/260-sdnotify.bats:189]: The way I read this feature request is to have podman wait until the service inside the container has finished its startup sequence before notifying systemd that it is ready. As such, I would assume that the startup health checks are the ones driving it.
And yes, the idea is to separate between "succeeded/failed to start" and runtime failures. 
Having said that, if waiting for only the startup is not supported, this PR can be merged

Comment (vrothberg) [reply test/system/260-sdnotify.bats:189]: Thanks for confirming, @ygalblum!  Let's chat about how to support startup HCs in tomorrow's meeting.
